crates/external_services/src/grpc_client/dynamic_routing.rs

```diff
@@ -0,0 +1,220 @@
+            let channel = tonic::transport::Endpoint::new(uri)?.connect().await?;
+            Some(SuccessRateCalculatorClient::new(channel))
+        } else {
+            None
+        };
+        Ok(RoutingStrategy {
+            success_rate_client,
+        })
+    }
+}
+
+/// The trait Success Based Dynamic Routing would have the functions required to support the calculation and updation window
+#[async_trait::async_trait]
+pub trait SuccessBasedDynamicRouting: dyn_clone::DynClone + Send + Sync {
+    /// To calculate the success rate for the list of chosen connectors
+    async fn calculate_success_rate(
+        &self,
+        id: id_type::ProfileId,
+        dynamic_routing_config: SuccessBasedRoutingConfig,
+        label_input: Vec<RoutableConnectorChoice>,
+    ) -> DynamicRoutingResult<CalSuccessRateResponse>;
+    /// To update the success rate with the given label
+    async fn update_success_rate(
+        &self,
+        id: id_type::ProfileId,
+        dynamic_routing_config: SuccessBasedRoutingConfig,
+        response: Vec<RoutableConnectorChoiceWithStatus>,
+    ) -> DynamicRoutingResult<UpdateSuccessRateWindowResponse>;
+}
+
+#[async_trait::async_trait]
+impl SuccessBasedDynamicRouting for SuccessRateCalculatorClient<Channel> {
+    async fn calculate_success_rate(
+        &self,
+        id: id_type::ProfileId,
+        dynamic_routing_config: SuccessBasedRoutingConfig,
+        label_input: Vec<RoutableConnectorChoice>,
+    ) -> DynamicRoutingResult<CalSuccessRateResponse> {
+        let params = dynamic_routing_config
+            .params
+            .map(|vec| {
+                vec.into_iter().fold(String::new(), |mut acc_vec, params| {
+                    if !acc_vec.is_empty() {
+                        acc_vec.push(':')
+                    }
+                    acc_vec.push_str(params.to_string().as_str());
+                    acc_vec
+                })
+            })
+            .get_required_value("Vector of params")
+            .change_context(DynamicRoutingError::MissingRequiredParamToBuildRequest)?;
+
+        let labels = label_input
+            .into_iter()
+            .map(|conn_choice| conn_choice.to_string())
+            .collect::<Vec<_>>();
+
+        let config = dynamic_routing_config.config.map(ForeignFrom::foreign_from);
+
+        let request = tonic::Request::new(CalSuccessRateRequest {
+            id: id.get_string_repr().to_owned(),
+            params,
+            labels,
+            config,
+        });
+
+        let mut client = self.clone();
+
+        let response = client
+            .fetch_success_rate(request)
+            .await
+            .change_context(DynamicRoutingError::GrpcServerResponseFailure)?
+            .into_inner();
+
+        Ok(response)
+    }
+
+    async fn update_success_rate(
+        &self,
+        id: id_type::ProfileId,
+        dynamic_routing_config: SuccessBasedRoutingConfig,
```

Comment:
```suggestion
        success_based_routing_config: SuccessBasedRoutingConfig,
```
Let's try to change this in every possible places so that the model will be prepared beforehand, for other variants of Dynamic Routing.